File: PythonSolutions/Neetcode150/graph_problems/210_course-schedule-2.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preMap = defaultdict(list)

        # Map course -> pre-req as the adjacency list
            # Think about how it backtracks when call stack pops
            # We visit courses as deep as possible after visiting it pops, and we add that deepest pre-req node first
        for crs, pre in prerequisites:
            preMap[crs].append(pre)
        
        visit = set()
        done = set()
        #! Sets aren't ordered in Python so we use a list
        order = []

        def dfs(crs):
            if crs in visit:
                return False
            if crs in done:
                return True
            
            visit.add(crs)

            # visit the pre-requisites as deep as possible
            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            
            # We backtrack here. Pre-requisites will be added first in the order.
            visit.remove(crs)
            done.add(crs)
            order.append(crs)

            return True
        
        # Visit all courses in course schedule
        for crs in range(numCourses):
            if not dfs(crs):
                return []
        
        return order

File: TypescriptSolutions/Neetcode150/graph_grid_problems/210_course-schedule-2.ts
export {};
const findOrder = (numCourses: number, prerequisites: number[][]): number[] => {
    const preMap: Record<number, number[]> = {};

    for (let i = 0; i < numCourses; i++) preMap[i] = [];

    // Courses -> Prerequisites
    for (const [crs, pre] of prerequisites) {
        preMap[crs].push(pre);
    }

    const visit = new Set<number>();
    const completed = new Set<number>();

    const dfs = (crs: number): boolean => {
        if (visit.has(crs)) return false;
        if (completed.has(crs)) return true;

        visit.add(crs);

        for (const pre of preMap[crs]) {
            if (!dfs(pre)) return false;
        }

        completed.add(crs);
        visit.delete(crs);

        return true;
    };

    for (let i = 0; i < numCourses; i++) {
        if (!dfs(i)) return [];
    }

    return [...completed];
};